### PR TITLE
[components] @component and local scope

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This PR proposes changing our registration mechanism.

First, instead of an abstract classmethod to determine the registered name, use a decorator `@component`. This plays nicer with inheritance (you do not want to inherit component keys)

`defs.yml` files before this PR had a `component_type` key that indexed into the component library. This works fine for pre-defined components that you don't need to customize. However you often want to specialize execution. In non-component integrations, this means changing the body of an asset-decorated function.

So that end it adds a new component-local syntax fro component type, and reuse that same decorator. The rule is that if the component type is prefixed with a "." we will dynamically load all the python in the folder until a registered component is found.

This works differently. Instead:

* The parent component class defines an `execute` method which has a default implementation
* A child component can override that to customize behavior.

So in the example here, we want to customize the behavior of the sling integration.

```python
@component(name="debug_sling_replication")
class DebugSlingReplicationComponent(SlingReplicationComponent):
    def execute(self, context: AssetExecutionContext, sling: SlingResource) -> Iterator:
        return sling.replicate(context=context, debug=True)
```

And in the `defs.yml` file

```yaml
component_type: .debug_sling_replication
component_params:
  connections:
    - name: snowflake
      type: snowflake
```

The prefix of "." tells the component system to look for a local module instead of going to the component registry.

## How I Tested These Changes

Manual

### CODE REVIEW

**Comment by schrockn** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26301?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26301** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26301?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26301?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26298** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26298?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26297** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26297?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: examples/experimental/dagster-components/dagster_components/core/component_defs_builder.py [hunk 1]
```diff
@@ -19,13 +22,48 @@
 if TYPE_CHECKING:
     from dagster._core.definitions.definitions_class import Definitions
 
+import importlib.util
+import inspect
+
+
+def load_module_from_path(module_name, path) -> ModuleType:
```
Comment (schrockn) [inline examples/experimental/dagster-components/dagster_components/core/component_defs_builder.py:29]: Generated this from chatGPT. Similar to `import_module_from_path` in dagster._seven but does not interact with `sys.modules`.

File: examples/experimental/dagster-components/dagster_components/impls/sling_replication.py [hunk 2]
```diff
@@ -46,3 +49,8 @@ def generate_files(cls) -> None:
                 {"source": {}, "target": {}, "streams": {}},
                 f,
             )
+
+    def execute(
+        self, context: AssetExecutionContext, sling: SlingResource
+    ) -> Iterator[Union[AssetMaterialization, MaterializeResult]]:
+        yield from sling.replicate(context=context)
```
Comment (schrockn) [inline examples/experimental/dagster-components/dagster_components/impls/sling_replication.py:56]: we would eventually make this pattern more generic

File: examples/experimental/dagster-components/dagster_components_tests/code_locations/custom_sling_location/custom_sling_location/components/debug_sling_component/defs.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+component_type: ./debug_sling_component
```
Comment (OwenKephart) [inline examples/experimental/dagster-components/dagster_components_tests/code_locations/custom_sling_location/custom_sling_location/components/debug_sling_component/defs.yml:1]: not a huge fan of this `./` syntax -- feels like a weird mix of filesystem path and not filesystem path (bc it doesn't have the `.py`)

what if this was:

```python
component_type:
  local: debug_sling_component
```

**Review by OwenKephart — COMMENTED** [review]
mostly on board, waffling a bit internally on the decorator bit but all in all I think it's not too bad to have to be explicit about marking one of these things (vs. us "intelligently" finding the right attribute in the file)

have a comment about the syntax of the component type

File: examples/experimental/dagster-components/dagster_components_tests/code_locations/custom_sling_location/custom_sling_location/components/debug_sling_component/defs.yml (hunk 3 above)
Comment (schrockn) [reply examples/experimental/dagster-components/dagster_components_tests/code_locations/custom_sling_location/custom_sling_location/components/debug_sling_component/defs.yml:1]: Yeah neither of these is particularly satisfactory right now. Will come up with some alts.

**Review by schrockn — COMMENTED** [review]
Updated this @OwenKephart. Went with a bare "." to mimic python module syntax.

Also @smackesey this proposes using a decorator to annotate components for registration, mostly to avoid awkwardness with inheritance.

**Review by schrockn — COMMENTED** [review]
Feedback from live convo: `registered_component` --> `component`.

File: python_modules/libraries/dagster-components/dagster_components/core/component.py [hunk 4]
```diff
@@ -89,3 +85,49 @@ def for_test(
...
             resources=resources or {}, registry=registry or ComponentRegistry.empty()
         )
+
+
+COMPONENT_REGISTRY_KEY_ATTR = "__dagster_component_registry_key"
+
+
+def component(
+    cls_or_key: Optional[Union[str, Type[Component]]] = None,
+    component_registry_key: Optional[str] = None,
```
Comment (smackesey) [inline python_modules/libraries/dagster-components/dagster_components/core/component.py]: Can we make the key keyword-only:

```
@component(key=...)
```

File: python_modules/libraries/dagster-components/dagster_components/core/component_defs_builder.py [hunk 5]
```diff
@@ -36,6 +54,34 @@ def build_components_from_decl_node(
...
+
+            module = load_module_from_path(module_name, decl_node.path / f"{module_name}.py")
+
+            for _name, obj in inspect.getmembers(module, inspect.isclass):
+                assert isinstance(obj, Type)
+                if (
+                    is_registered_component(obj)
+                    and component_registry_key_of(obj) == component_registry_key
+                ):
+                    return obj
```
Comment (smackesey) [inline python_modules/libraries/dagster-components/dagster_components/core/component_defs_builder.py:73]: Perhaps best left for a follow-up but I am partial to avoiding `load_module_from_path` and instead requiring the containing code location to be installed in the python environment

File: python_modules/libraries/dagster-components/dagster_components/core/component.py [hunk 6]
```diff
@@ -89,3 +85,49 @@ def for_test(
...
+        check.inst_param(cls, "cls", type)
+        setattr(cls, COMPONENT_REGISTRY_KEY_ATTR, component_registry_key or snakecase(cls.__name__))
+        return cls
+
+
+def is_registered_component(cls: Type) -> bool:
+    return hasattr(cls, COMPONENT_REGISTRY_KEY_ATTR)
+
+
+def component_registry_key_of(component_type: Type[Component]) -> str:
```
Comment (smackesey) [inline python_modules/libraries/dagster-components/dagster_components/core/component.py:121]: nit: fn name feels awkward-- `get_component_name`?

**Comment by schrockn** [comment]
Required keyword arg for `component` and changed name to `name`.

**Review by smackesey — APPROVED** [review]
Some comments but good to merge when you're ready

File: python_modules/libraries/dagster-components/dagster_components/core/component_defs_builder.py (hunk 5 above)
Comment (schrockn) [reply python_modules/libraries/dagster-components/dagster_components/core/component_defs_builder.py:73]: Yeah let's do that in follow up

**Comment by schrockn** [comment]
### Merge activity

* **Dec 6, 3:02 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/26301).

**Pull request merged**